test/core/test_serde_dataclass.py

```diff
@@ -33,9 +35,31 @@ def __eventually__(self, cardinality):
             cardinality.set_default([1])
 
 
+class A(BaseModel):
+    pass
+
+
+class B(A):
+    pass
+
+
+@serde.dataclass
+class X:
+    a: A
+
+
 def test_dataclass():
     est = Estimator(prediction_length=7)
 
     assert est.prediction_length == 7
     assert est.context_length == 14
     assert est.cardinality == [1]
+
+
+def test_dataclass_subtype():
+    x = X(a=B())
+
+    assert isinstance(x.a, B)
+
+    x2 = serde.decode(serde.encode(x))
+    assert isinstance(x2.a, B)
```

Comment:
Yes, the problem is this:

```py
In [10]: from pydantic import BaseModel
    ...: 
    ...: 
    ...: class A(BaseModel):
    ...:     pass
    ...: 
    ...: class B(A):
    ...:     pass
    ...: 
    ...: 
    ...: class M(BaseModel):
    ...:     a: A
    ...: 
    ...: m = M(a=B())
    ...: 
    ...: m.dict()
Out[10]: {'a': {}}
```

When deserialising, we want to reconstruct some `A` from `{}`, which then returns an `A` instead of the `B`.